packages/adapters/watcher/src/verifiers/proposal.ts

```diff
@@ -0,0 +1,249 @@
+      const eventFilter = rootManager.filters.AggregateRootProposed();
+      const events = await rootManager.queryFilter(eventFilter);
+
+      // sort events by block number to get the latest proposed root
+      const [latestProposedRootEvent] = events.sort((a, b) => b.blockNumber - a.blockNumber);
+
+      if (!latestProposedRootEvent?.args) {
+        throw new Error(`No valid events found for root proposal`);
+      }
+
+      return latestProposedRootEvent.args as unknown as ProposedData;
+    } catch (error: any) {
+      this.context.logger.error(
+        "Failed when getting the latest proposed aggregate root",
+        requestContext,
+        methodContext,
+        jsonifyError(error as Error),
+        {
+          domain: this.hubDomain,
+          chainId: hubDomainChainId,
+          rootManager: rootManager.address,
+        },
+      );
+      throw error;
+    }
+  }
+
+  /**
+   * @notice Get the virtual aggregate root.
+   * @param proposedData - ProposedData the data returned by the last AggregateRootProposed event.
+   * @param spokeConnectors - List of every spoke connector instance for every domain used in the propose.
+   * @returns string representing the virtual root.
+   */
+  public async getSnapshotRoots(
+    proposedData: ProposedData,
+    spokeConnectors: { address: string; abi: any }[],
+  ): Promise<string[]> {
+    // Get the snapshot roots from domains. We parallelize the calls to the spoke connectors.
+    const snapshotRoots = await Promise.all(
+      spokeConnectors.map(async (spokeConnector, i) => {
+        // spokeConnectors are ordered by domain, so we can use the index to get the domain
+        const domain = +proposedData.domains[i];
+        const spokeConnectorInterface = new utils.Interface(spokeConnector.abi as string[]);
+        let snapshotRoot = await this.getSnapshotRootForDomain(
+          proposedData.snapshotId,
+          spokeConnector.address,
+          spokeConnectorInterface,
+          domain,
+        );
+
+        // If the snapshot root is empty, we need to get the root from `getOutboundRoot` method
+        if (snapshotRoot == EMPTY_ROOT) {
+          snapshotRoot = await this.getOutboundRoot(spokeConnector.address, spokeConnectorInterface, domain);
+        }
+
+        // push the snapshot root to the array
+        return snapshotRoot;
+      }),
+    );
+
+    return snapshotRoots;
+  }
+
+  /**
+   * @notice Uses the batching method to get the latest aggregate root from the MerkleTreeManager and inserts each snapshot root of each domain to get the correct virtual aggregate root
+   * @param leaves - Array of the snapshoot roots of each domain. Its order must follow the order of the domains emitted in the AggregateRootProposed event.
+   * @return virtualRoot The virtual aggregate root computed after the insertion of the leaves
+   */
+  private async insertVirtualLeaves(leaves: string[]): Promise<string> {
```

Comment:
is there anyway this could get out of sync with what is maintained off-chain since the root merkle tree doesn't get updated each time a new set of leaves is proposed?